src/foam/nanos/crunch/ReputDependentUCJs.js

```diff
@@ -0,0 +1,106 @@
+/**
+ * @license
+ * Copyright 2020 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.nanos.crunch',
+  name: 'ReputDependentUCJs',
+
+  documentation: `When a ucj goes into a certain status, try to reput its dependents
+  `,
+
+  implements: [
+    'foam.nanos.ruler.RuleAction'
+  ],
+
+  javaImports: [
+    'foam.core.ContextAgent',
+    'foam.core.X',
+    'foam.dao.ArraySink',
+    'foam.dao.DAO',
+    'java.util.ArrayList',
+    'java.util.List',
+    'static foam.mlang.MLang.*'
+  ],
+
+  methods: [
+    {
+      name: 'applyAction',
+      javaCode: `
+        agency.submit(x, new ContextAgent() {
+          @Override
+          public void execute(X x) {
+            DAO userCapabilityJunctionDAO = (DAO) x.get("userCapabilityJunctionDAO");
+            UserCapabilityJunction ucj = (UserCapabilityJunction) obj;
+            UserCapabilityJunction old = (UserCapabilityJunction) userCapabilityJunctionDAO.find(AND(
+              EQ(UserCapabilityJunction.SOURCE_ID, ucj.getSourceId()),
+              EQ(UserCapabilityJunction.TARGET_ID, ucj.getTargetId())
+            ));
+            CapabilityJunctionStatus status = ucj.getStatus();
+
+            boolean isInvalidate = ucj.getStatus() != CapabilityJunctionStatus.GRANTED || ucj.getStatus() != CapabilityJunctionStatus.GRACE_PERIOD;
+            
+            DAO filteredUserCapabilityJunctionDAO = (DAO) userCapabilityJunctionDAO.where((EQ(UserCapabilityJunction.SOURCE_ID, ucj.getSourceId())));
+            DAO filteredPrerequisiteCapabilityJunctionDAO = (DAO) ((DAO) x.get("prerequisiteCapabilityJunctionDAO"))
+              .where(EQ(CapabilityCapabilityJunction.TARGET_ID, ucj.getTargetId()));
+            
+            List<CapabilityCapabilityJunction> ccjs = ((ArraySink) filteredPrerequisiteCapabilityJunctionDAO
+              .select(new ArraySink()))
+              .getArray();
+
+            List<UserCapabilityJunction> ucjsToReput = new ArrayList<UserCapabilityJunction>();
+
+            for ( CapabilityCapabilityJunction ccj : ccjs ) {
+              UserCapabilityJunction ucjToReput = (UserCapabilityJunction) filteredUserCapabilityJunctionDAO
+                .find(EQ(UserCapabilityJunction.TARGET_ID, ccj.getSourceId()));
+              if ( ucjToReput != null ) ucjsToReput.add(ucjToReput);
+            }
+
+            for ( UserCapabilityJunction ucjToReput : ucjsToReput ) {
+              if ( isInvalidate ) ucjToReput.setStatus(cascadeInvalidateStatus(x, ucjToReput, status));
```

Comment:
if a dependent ucj is being reput as a result of this rule, we can safely assume that its data has not changed since the last time it was put, so it made more sense to me to not have it go through validation checks again

Reply:
So ya re-reading my comment - I was in error. This does work. Thanks for clarifying though.